### ISSUE TITLE
Format for input data

### ISSUE THREAD

**Issue by goldblum** [facebookresearch/ImageNet-Adversarial-Training#11]
Hello,

I'm trying to run your inference-example.py file, and I'm confused about how the input data should be formatted. You have some normalization in the inference-example.py file, but also in when you build the graph. Should I be plugging in a uint8 array, or some type of normalized array. I'm trying to use your model with my own modified dataset.

**Comment by cihangxie** [comment]
For the input, you just need to provide valid RGB images. The inference-example.py has done every pre-processing that you need for getting right predictions.

**Comment by goldblum** [comment]
Okay, we're using our own dataset and dataloader, so we won't be using the stock inference code. I think we've figured it out, but just to confirm, we load numpy arrays of uint8 entries (RGB not BGR), divide by 127.5, subtract 1, and feed this into the model so that the data fed into the model has entries between -1 and 1. Is this correct? For some reason, we're only getting natural accuracy of ~69%. Or do we need to normalize the data first (set mean=0 and std=1)?  If so, do we just feed normalized data straight into the model?  When we feed normalized data into the model, we get terrible accuracy.

Also, we're more familiar with pytorch. Do we have to do anything to your model to have it not update the weights for inference on a large set of data (i.e. run in inference mode)? 

Finally, does your inference script load batch norm running mean and std dev? Our results seemed to be changing over time.

**Comment by cihangxie** [comment]
The right preprocessing procedure:
(1) get image in the BGR format (so you need to rotate the channel axis to make your RGB to BGR)
(2) resize it to 224x224 (if your images come from ImageNet, we recommend to the standard pre-processing --- resize shortest edge to 256 + center crop 224)
(3) image normalization --- divide by 127.5, and then subtract 1
(4) transpose the image from NHWC to NCHW

Then you can feed obtained images to the network.

Accuracy around 69% seems correctly. You can refer the README for the performance that we reported.

If you run the code in the test mode, the parameters will not be updated and the saved running mean/variance will be used. 

For reference, we use `with TowerContext('', is_training=False):` to make sure the graph is operated under the test mode.

**Comment by ppwwyyxx** [comment]
Clarification: the preprocessing @cihangxie mentioned above produces the expected input format for the `model.get_logits(image)` function, which is also used by `inference-example.py`.

If you build your graph by `model.build_graph(image, label)` instead, it expects only (1) and (2) as the preprocessing. (3) and (4) are performed inside the `build_graph` function already.

**Comment by goldblum** [comment]
Thanks for your help.  Everything seems to be working.
